File: 백준/Gold/1074. Z/Z.java
import java.util.*;
import java.lang.*;
import java.io.*;

// The main method must be in a class named "Main".
class Main {
    static int N, R, C;
    static int answer = 0;
    static int[] drow = new int[] {0, 0, 1, 1};
    static int[] dcol = new int[] {0, 1, 0, 1};
    
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());
        R = Integer.parseInt(st.nextToken());
        C = Integer.parseInt(st.nextToken());

        int num = (int)Math.pow(2, N);
        recursion(0, 0, num, num);
        System.out.println(answer);
    }

    static void recursion(int start_row, int start_col, int end_row, int end_col) {
        if (start_row == R && start_col == C) {
            return;
        }
 
        int len = end_row - start_row;
        int half = len / 2;

        
        int total = 0;  
        for (int i = 0; i < 4; i++) {
            int new_start_row = start_row + drow[i] * half;
            int new_start_col = start_col + dcol[i] * half;
            int new_end_row = start_row + half + drow[i] * half;
            int new_end_col = start_col + half + dcol[i] * half;

            if (new_start_row <= R && new_end_row > R &&
               new_start_col <= C && new_end_col > C) {
                recursion(new_start_row, new_start_col, new_end_row, new_end_col);
                break;
               }

            total += half * half;
        }

        answer += total;
    }
}

File: 백준/Gold/1074. Z/Z.js
const fs = require('fs');
const input = fs.readFileSync('/dev/stdin').toString().split(' ');

const [ N, R, C ] = input.map(Number);

let ans = 0;

function recursion(start_row, start_col, end_row, end_col) {
    if (start_row === R && start_col === C) return;
    
    const length = end_row - start_row;
    const half = length / 2;
    
    //1 
    if (R >= start_row && R < start_row + half &&
       C >= start_col && C < start_col + half) {
        recursion(start_row, start_col, start_row + half, start_col + half);
    }
    else if (R >= start_row && R < start_row + half &&
            C >= start_col + half && C < end_col) {
        ans += (half * half);
        recursion(start_row, start_col + half, start_row + half, end_col);
    }
    else if (R >= start_row + half && R < end_row &&
            C >= start_col && C < start_col + half){
        ans += (half * half) * 2;
        recursion(start_row + half, start_col, end_row, start_col + half);    
    }
    else {
        ans += (half * half) * 3;
        recursion(start_row + half, start_col + half, end_row, end_col);   
    }
       
}

recursion(0, 0, 2**N, 2**N);

console.log(ans);